dotultimate: init at 2025.2.4

js6pak opened the pull request:
https://www.jetbrains.com/dotmemory
https://www.jetbrains.com/dottrace

Also adds a [extract-dotnet-resources](https://github.com/NixOS/nixpkgs/commit/ad4dce638364b528646ea0905607d331931917c1) utility to extract the icon from .NET assemblies.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jtojnar commented on pkgs/development/tools/misc/dotultimate/default.nix:
```diff
@@ -0,0 +1,241 @@
+          local download_sha256=$(curl -s "$(jq -r ".checksumLink" <<< "$download")" | cut -d ' ' -f1)
+
+          update-source-version "$UPDATE_NIX_ATTR_PATH" "$version" "$download_sha256" "$download_url" \
+            --system="$system" --ignore-same-version
```
`--system="$system"` does not really work – for example, it is not generally possible to use it to update Darwin source on Linux. I really need to remove it.

Instead, I would recommend exposing the sources in `passthru` and then updating it as attributes, for example: https://github.com/NixOS/nixpkgs/blob/ad8d476b53bd39c93b33d6542eade24230114ac8/pkgs/by-name/gk/gk-cli/package.nix#L45-L115

js6pak replied:
> `--system="$system"` does not really work

It's [used quite a lot in nixpkgs](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs%20%22--system%3D%22&type=code), are all of those wrong? 

> it is not generally possible to use it to update Darwin source on Linux. 

It's working for me, so I'm curious when would it not work? (except IFD I guess)
 

I'm getting both the url and hash from the updater api anyway so the way to go might be just generating a json file? Or is there a way to modify values in a nix file more semantically without relying on sed?

jtojnar replied:
Hmm, might be the case that it works when a hash is specified, since that likely only needs eval. When hash is not provided, Nix will need to build the source derivation so that will fail without compatible builders.

jtojnar commented on pkgs/common-updater/scripts/update-source-version:
```diff
@@ -229,12 +229,14 @@ if [[ -n "$newUrl" ]]; then
         die "Couldn't evaluate source url from '$attr.$sourceKey'!"
     fi
 
-    # Escape regex metacharacter that are allowed in store path names
```
The URL check is there intentionally to detect situation where version changes but update script does not update the URL. Though maybe we could skip the check with `ignoreSameVersion` when explicit URL is provided.

js6pak replied:
> Though maybe we could skip the check with `ignoreSameVersion` when explicit URL is provided.

So `if [[ -z "$ignoreSameVersion" || "$oldUrl" != "$newUrl" ]]; then`?

jtojnar replied:
Actually, disregard that as well – your change makes sense as it is. The check I had in mind is the comparison next to version.

jtojnar commented:
common-updater-scripts change looks good to me, would be nice if you included some context how the feature is useful in the commit message.

jtojnar commented:
Thanks. Though I was more interested in why one would want to perform an update to the same URL.

The motivation for `--ignore-same-version` was also updating hashes for multiple attributes (e.g. `src` and `cargoHash`): https://github.com/NixOS/nixpkgs/commit/4627e4849e6df96d1b18b14144d39256e606387e

It is not clear to me what would the use case be here. If the URL were the same then the hash would be the same and there would be nothing to update. The only scenarios I was able to come up with were pretty unlikely IMO:

- When source for some of the platforms stayed behind – though then the `version`  attribute would no longer be correct for those.
- There are multiple FODs with the same source URL but different hash e.g. because of different `postFetch`.

js6pak commented:
> It is not clear to me what would the use case be here

For this package it's really so the update script doesn't crash when there is no update
```
update-source-version: error: Failed to replace source URL 'https://download.jetbrains.com/resharper/dotUltimate.2024.2.6/JetBrains.dotTrace.macos-arm64.2024.2.6.dmg' to 'https://download.jetbrains.com/resharper/dotUltimate.2024.2.6/JetBrains.dotTrace.macos-arm64.2024.2.6.dmg' in 'dotultimate.dottrace'!
```

FliegendeWurst commented on pkgs/top-level/all-packages.nix:
```diff
@@ -38874,4 +38875,6 @@ with pkgs;
     fltk = fltk13;
   };
+
+  dotultimate = recurseIntoAttrs (callPackages ../development/tools/misc/dotultimate {});
```
I think these should go in pkgs/by-name

js6pak replied:
[`pkgs/by-name/README.md`](https://github.com/NixOS/nixpkgs/blob/799661e93525c7fa2c9acf645cafe82bab3ee261/pkgs/by-name/README.md#limitations) seems to still suggest using the category hierarchy for package sets.

FliegendeWurst replied:
Hmm... it isn't really a "package set". Just some related packages.

FliegendeWurst replied:
You would need to add a dotultimate package, maybe just a symlink join of the two tools if that makes sense here. Then you can add the other packages in the same directory and reference them in passthru.

Users could choose between dotultimate, dotultimate.dotmemory, dotultimate.dottrace then.

js6pak committed:
dotultimate: init at 2025.2.4

js6pak committed:
common-updater-scripts: update url only when it changed

This mimics the `"$oldVersion" != "$newVersion"` check which is necessary when you use `--ignore-same-version` and url didn't change.

# Conflicts:
#	pkgs/common-updater/scripts/update-source-version

# Conflicts:
#	pkgs/common-updater/scripts/update-source-version

js6pak committed:
extract-dotnet-resources: init
